File: Examenes-EDD-1/2016-Febrero/Haskell/DataStructures/Graph/EulerianCycle.hs
-------------------------------------------------------------------------------
-- Student's name: Paco Molina Cuenca
-- Student's group: Doble Grado Matemáticas e Ingeniería Informática
--
-- Data Structures. Escuela Térnica Superior de Ingenieria Informática. UMA.
-------------------------------------------------------------------------------

module DataStructures.Graph.EulerianCycle (isEulerian, remove, extractCycle, connectCycles, vertexInCommon, eulerianCycle) where

import DataStructures.Graph.Graph
import Data.List

--H.1)
isEulerian :: Eq a => Graph a -> Bool
isEulerian g = foldr (&&) True (map (\x -> (degree g x) `mod` 2 == 0) (vertices g))

-- H.2)
remove :: (Eq a) => Graph a -> (a,a) -> Graph a
remove g (v1,v2) = deleteIsolated (deleteEdge g (v1,v2)) (vertices g)
    where 
        deleteIsolated :: (Eq a) => Graph a -> [a] -> Graph a
        deleteIsolated g [] = g
        deleteIsolated g (v:vs) = 
            if degree g v == 0
                then deleteIsolated (deleteVertex g v) vs 
                else deleteIsolated g vs

-- H.3)
extractCycle :: (Eq a) => Graph a -> a -> (Graph a, Path a)
extractCycle g v0 = extractCycle' g v0 [v0]
    where 
        extractCycle' g v path 
            | v == v0 && length path > 1 = (g, reverse path)
            | otherwise = 
                let u = head (successors g v)
                    g' = remove g (v,u)
                    path' = u : path
                in extractCycle' g' u path'
   
-- H.4)
connectCycles :: (Eq a) => Path a -> Path a -> Path a
connectCycles [] ys = ys 
connectCycles (x:xs) (y:ys)
    | x == y =  x:ys ++ xs
    | otherwise = x : connectCycles xs (y:ys)

-- H.5)
vertexInCommon :: Eq a => Graph a -> Path a -> a
vertexInCommon g (x:xs) =
    if x `elem` (vertices g)
        then x 
        else vertexInCommon g xs

-- H.6) 
eulerianCycle :: Eq a => Graph a -> Path a
eulerianCycle g 
    | not(isEulerian g) = error "eulerianCycle on not eulerian graph"
    | otherwise = eulerianCycle' g verticeArbitrario []
    where 
        verticeArbitrario = head (vertices g)

        eulerianCycle' g v path
            | isEmpty g = path
            | otherwise = 
                let (g', ciclo) = extractCycle g v
                in eulerianCycle' g' (vertexInCommon g' ciclo) (connectCycles path ciclo)

File: Examenes-EDD-1/2016-Febrero/Java/dataStructures/graph/EulerianCycle.java
/**
 * Student's name: Paco Molina Cuenca
 * Student's group: Doble Grado Matemáticas e Ingeniería Informática
 * Data Structures. E.T.S.I. Informática. UMA.
 */

package dataStructures.graph;
import dataStructures.list.*;

public class EulerianCycle<V> {
    private List<V> eCycle;

    @SuppressWarnings("unchecked")
    public EulerianCycle(Graph<V> g) {
        Graph<V> graph = (Graph<V>) g.clone();
        eCycle = eulerianCycle(graph);
    }

    public boolean isEulerian() {
        return eCycle != null;
    }

    public List<V> eulerianCycle() {
        return eCycle;
    }

    // J.1
    private static <V> boolean isEulerian(Graph<V> g) {
        boolean esEuleriano = true;

        for(V v : g.vertices()){
            if(g.degree(v) % 2 != 0)
                esEuleriano = false;
        }

        return esEuleriano;
    }

    // J.2
    private static <V> void remove(Graph<V> g, V v, V u) {
        g.deleteEdge(v, u);

        for(V w : g.vertices()){
            if(g.degree(w) == 0)
                g.deleteVertex(w);
        }
    }

    // J.3
    private static <V> List<V> extractCycle(Graph<V> g, V v0) {
        List<V> ciclo = new LinkedList<>();
        ciclo.append(v0);
        V v = v0;
        V u;

        while(!(v.equals(v0) && ciclo.size() > 1)){
            u = sucesor(g, v);
            remove(g, v, u);
            ciclo.append(u);
            v = u;
        }

        return ciclo;
    }

    private static <V> V sucesor(Graph<V> g, V u){
        boolean esPrimerSucesor = false;
        V primerSucesor = null;

        for(V sucesor : g.successors(u)){
            if(!esPrimerSucesor){
                primerSucesor = sucesor;
                esPrimerSucesor = true;
            }
        }

        return primerSucesor;
    }

    // J.4
    private static <V> void connectCycles(List<V> xs, List<V> ys) {
        if(xs.isEmpty()){
            for(int i = 0; i < ys.size(); i++)
                xs.append(ys.get(i));
        
        } else{
            List<V> res = new LinkedList<>();

            int i = 0;
            V elem = ys.get(0);

            while(!xs.get(i).equals(elem)){
                res.append(xs.get(i));
                i++;
            }

            for(V v : ys)
                res.append(v);
            
            for(i = i + 1; i < xs.size(); i++)
                res.append(xs.get(i));
            
            for(int j = 0; j < xs.size(); j++)
                xs.remove(j);
            
            for(int j = 0; j < res.size(); j++)
                xs.append(res.get(j));
        }
    }

    // J.5
    private static <V> V vertexInCommon(Graph<V> g, List<V> cycle) {
        boolean esVerticeComun = false;
        V verticeComun = null;

        for(V v : cycle){
            if(!esVerticeComun && g.vertices().isElem(v)){
                esVerticeComun = true;
                verticeComun = v;
            }
        }

        return verticeComun;
    }

    // J.6
    private static <V> List<V> eulerianCycle(Graph<V> g) {
        if(!isEulerian(g)){
          return null;
        
        } else{
            List<V> path = new LinkedList<>();
            V v = verticeArbitrario(g);
            List<V> ciclo;

            while(!g.isEmpty()){
                ciclo = extractCycle(g, v);
                v = vertexInCommon(g, ciclo);
                connectCycles(path, ciclo);
            }

            return path;
        }
    }

    private static <V> V verticeArbitrario(Graph<V> g){
        V vertice = null;
        boolean esPrimerVertice = false;

        for(V v : g.vertices()){
            if(!esPrimerVertice){
                esPrimerVertice = true;
                vertice = v;
            }
        }

        return vertice;
    }
}
